README.md
[![Build Status](https://travis-ci.org/andrba/aba.svg?branch=master)](https://travis-ci.org/andrba/aba) [![Code Climate](https://codeclimate.com/github/andrba/aba/badges/gpa.svg)](https://codeclimate.com/github/andrba/aba)

# Aba

Generates ABA (Australian Banking Association) file format output

## Usage

```ruby
require 'aba'

# Initialise ABA
aba = Aba.batch(
  bsb: "123-345", # Optional (Not required by NAB)
  financial_institution: "WPC",
  user_name: "John Doe",
  user_id: "466364",
  description: "Payroll",
  process_at: Time.now.strftime("%d%m%y")
)

# Add transactions...
10.times do
  aba.add_transaction(
    {
      bsb: "342-342",
      account_number: "3244654",
      amount: 10000, # Amount in cents
      account_name: "John Doe",
      transaction_code: 53,
      lodgement_reference: "R435564",
      trace_bsb: "453-543",
      trace_account_number: "45656733",
      name_of_remitter: "Remitter"
    }
  )
end

# ...or add returns
10.times do
  aba.add_return(
    {
        bsb: '453-543',
        account_number: '45656733',
        amount: 10000,
        account_name: 'John Doe',
        transaction_code: 53,
        lodgement_reference: 'R435564',
        trace_bsb: '342-342',
        trace_account_number: '3244654',
        name_of_remitter: 'Remitter',
        return_code: 8, 
        original_user_id: 654321,
        original_processing_day: 12,
    }
  )
end

puts aba.to_s # View output
File.write("/Users/me/dd_#{Time.now.to_i}.aba", aba.to_s) # or write output to file
```

There are a few ways to create a complete set of ABA data:

```ruby
# Transactions added to the defined ABA object variable
aba = Aba.batch financial_institution: 'ANZ', user_name: 'Joe Blow', user_id: 123456, process_at: 200615
aba.add_transaction bsb: '123-456', account_number: '000-123-456', amount: 50000, transaction_code: 50
aba.add_transaction bsb: '456-789', account_number: '123-456-789', amount: 10000, transaction_code: 13

# Transactions passed individually inside a block
aba = Aba.batch financial_institution: 'ANZ', user_name: 'Joe Blow', user_id: 123456, process_at: 200615 do |a|
  a.add_transaction bsb: '123-456', account_number: '000-123-456', amount: 50000, transaction_code: 50
  a.add_transaction bsb: '456-789', account_number: '123-456-789', amount: 10000, transaction_code: 13
end

# Transactions as an array passed to the second param of Aba.batch
aba = Aba.batch(
  { financial_institution: 'ANZ', user_name: 'Joe Blow', user_id: 123456, process_at: 200615 },
  [
    { bsb: '123-456', account_number: '000-123-456', amount: 50000, transaction_code: 50 },
    { bsb: '456-789', account_number: '123-456-789', amount: 10000, transaction_code: 13 }
  ]
)
```

> **Note:** Positive (`n`) and negative (`-n`) values are now treated the same.
> e.g `5` and `-5` are both processed as `5`, without any signage.
> To differentiate between a debit and credit, use the correct [Transaction Code](https://github.com/andrba/aba/blob/58446f5b0ef822e9792e9399b4af647319b13515/lib/aba/transaction.rb#L106-L112)

Validation errors can be caught in several ways:

```ruby
# Create an ABA object with invalid character in the user_name
aba = Aba.batch(
  financial_institution: "ANZ",
  user_name: "Jøhn Doe",
  user_id: "123456",
  process_at: Time.now.strftime("%d%m%y")
)

# Add a transaction with a bad BSB
aba.add_transaction(
  bsb: "abc-123",
  account_number: "000123456"
)

# Is the data valid?
aba.valid?
# Returns: false

# Return a structured array of errors
puts aba.errors
# Returns:
# {:aba => ["user_name must not contain invalid characters"],
#  :transactions =>
#   {0 => ["bsb format is incorrect", "trace_bsb format is incorrect"]}}
```

Validation errors will stop parsing of the data to an ABA formatted string using
`to_s`. `aba.to_s` will raise a `RuntimeError` instead of returning output.


## Installation

Add this line to your application's Gemfile:

    gem 'aba'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install aba

## Contributing

1. Fork it ( https://github.com/[my-github-username]/aba/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# Changelog

## v1.0.2, 25 July 2022

### Added

- Enforce ABA validation:
  - Amount (in cents) cannot exceed 10 characters.
  - Total debit amount (in cents) cannot exceed 10 characters.
  - Total credit amount (in cents) cannot exceed 10 characters.

## v1.0.1, 11 January 2021

### BUG FIX

Account numbers/alphas validation should be case insensitive

## v1.0.0, 12 July 2020

### BREAKING CHANGES

- Positive (`n`) and negative (`-n`) values are now treated the same. e.g `5`
  and `-5` are both processed as `5`, without any signage. To differentiate
  between a debit and credit, use the correct [Transaction
  Code](https://github.com/andrba/aba/blob/58446f5b0ef822e9792e9399b4af647319b13515/lib/aba/transaction.rb#L106-L112)
- Removed default values for transactions records to avoid generation of
  potentially incorrect records. Safety first!
- Minimum Ruby version is now 2.5

### NEW FEATURE

- You can now add a *return* record to be used when you'd like to return a
  credit or a debit to another financial institution (OFI).


spec/aba_spec.rb
# encoding: UTF-8

require "spec_helper"

describe Aba do
  let(:aba) { Aba.new(financial_institution: "WPC", user_name: "John Doe",
                      user_id: "466364", description: "Payroll", process_at: "190615") }
  let(:transaction_values) { [30, -20] }
  let(:transactions) do
    transaction_values.map do |amount|
      Aba::Transaction.new(bsb: '342-342', account_number: '3244654', amount: amount,
      account_name: 'John Doe', transaction_code: 53,
      lodgement_reference: 'R435564', trace_bsb: '453-543',
      trace_account_number: '45656733', name_of_remitter: 'Remitter')
    end
  end
  subject { aba  }
  before { transactions.each { |trx| subject.add_transaction(trx) } }

  describe "#to_s" do

    context 'when descriptive record' do

      context 'without bsb' do
        it "should return a string containing the descriptive record without the bsb" do
          expect(subject.to_s).to include("0                 01WPC       John Doe                  466364Payroll     190615                                        \r\n")
        end
      end

      context 'with bsb' do
        before { subject.bsb = "123-345" }
        it "should return a string containing the descriptive record with the bsb" do
          expect(subject.to_s).to include("0123-345          01WPC       John Doe                  466364Payroll     190615                                        \r\n")
        end
      end
    end


    context 'when detail record' do

      it "should contain transactions records" do
        expect(subject.to_s).to include("1342-342  3244654 530000000030John Doe                        R435564           453-543 45656733Remitter        00000000\r\n")
        expect(subject.to_s).to include("1342-342  3244654 530000000020John Doe                        R435564           453-543 45656733Remitter        00000000\r\n")
      end
    end

    context 'when file total record' do

      context 'with unbalanced transactions' do
        it "should return a string wihere the net total is not zero" do
          expect(subject.to_s).to include("7999-999            000000001000000000300000000020                        000002                                        ")
        end
      end

      context 'with balanced transactions' do
        let(:transaction_values) { [30, 30, -60] }
        it "should return a string where the net total is zero" do
          expect(subject.to_s).to include("7999-999            000000000000000000600000000060                        000003                                        ")
        end
      end
    end
  end
end


spec/batch_spec.rb
# encoding: UTF-8

require "spec_helper"

describe Aba::Batch do
  let(:aba) { Aba::Batch.new(financial_institution: "WPC", user_name: "John Doe",
                      user_id: "466364", description: "Payroll", process_at: "190615") }
  let(:transaction_values) { [30, -20] }
  let(:transactions) do
    transaction_values.map do |amount|
      Aba::Transaction.new(bsb: '342-342', account_number: '3244654', amount: amount,
      account_name: 'John Doe', transaction_code: 53,
      lodgement_reference: 'R435564', trace_bsb: '453-543',
      trace_account_number: '45656733', name_of_remitter: 'Remitter')
    end
  end
  subject { aba  }
  before { transactions.each { |trx| subject.add_transaction(trx) } }

  describe "#to_s" do

    context 'when descriptive record' do

      context 'without bsb' do
        it "should return a string containing the descriptive record without the bsb" do
          expect(subject.to_s).to include("0                 01WPC       John Doe                  466364Payroll     190615                                        \r\n")
        end
      end

      context 'with bsb' do
        before { subject.bsb = "123-345" }
        it "should return a string containing the descriptive record with the bsb" do
          expect(subject.to_s).to include("0123-345          01WPC       John Doe                  466364Payroll     190615                                        \r\n")
        end
      end
    end


    context 'when detail record' do

      it "should contain transactions records" do
        expect(subject.to_s).to include("1342-342  3244654 530000000030John Doe                        R435564           453-543 45656733Remitter        00000000\r\n")
        expect(subject.to_s).to include("1342-342  3244654 530000000020John Doe                        R435564           453-543 45656733Remitter        00000000\r\n")
      end
    end

    context 'when file total record' do

      context 'with unbalanced transactions' do
        it "should return a string wihere the net total is not zero" do
          expect(subject.to_s).to include("7999-999            000000001000000000300000000020                        000002                                        ")
        end
      end

      context 'with balanced transactions' do
        let(:transaction_values) { [30, 30, -60] }
        it "should return a string where the net total is zero" do
          expect(subject.to_s).to include("7999-999            000000000000000000600000000060                        000003                                        ")
        end
      end
    end
  end
end


spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
Bundler.setup

require "aba"
require "pry"

RSpec.configure do |config|
  config.order = :random
  config.filter_run :focus
  config.run_all_when_everything_filtered = true
end


lib/aba/batch.rb
# frozen_string_literal: true

class Aba
  class Batch
    include Aba::Validations

    attr_accessor :bsb, :financial_institution, :user_name, :user_id, :description, :process_at, :entries

    # BSB
    validates_bsb         :bsb, allow_blank: true

    # Financial Institution
    validates_length      :financial_institution, 3

    # User Name
    validates_presence_of :user_name
    validates_max_length  :user_name, 26
    validates_becs        :user_name

    # User ID
    validates_presence_of :user_id
    validates_max_length  :user_id, 6
    validates_integer     :user_id, false

    # Description
    validates_max_length  :description, 12
    validates_becs        :description

    # Process at Date
    validates_length      :process_at, 6
    validates_integer     :process_at, false

    # Totals
    validates_max_length  :debit_total_amount, 10
    validates_max_length  :credit_total_amount, 10

    def initialize(attrs = {}, transactions = [])
      attrs.each do |key, value|
        send("#{key}=", value)
      end

      @entries = []

      unless transactions.nil? || transactions.empty?
        transactions.to_a.each do |t|
          self.add_transaction(t) unless t.nil? || t.empty?
        end
      end

      yield self if block_given?
    end

    def to_s
      raise RuntimeError, 'No entries present - add one using `add_transaction` or `add_return`' if entries.empty?
      raise RuntimeError, 'ABA data is invalid - check the contents of `errors`' unless valid?

      # Descriptive record
      output = "#{descriptive_record}\r\n"

      # Transactions records
      output += entries.map { |t| t.to_s }.join("\r\n")

      # Batch control record
      output += "\r\n#{batch_control_record}"
    end

    def add_transaction(attrs = {})
      add_entry(Aba::Transaction, attrs)
    end

    def add_return(attrs = {})
      add_entry(Aba::Return, attrs)
    end

    def transactions
      entries.select { |entry| entry.instance_of?(Aba::Transaction) }
    end

    def transactions_valid?
      !transactions.map { |t| t.valid? }.include?(false)
    end

    def valid?
      !has_errors? && !has_entry_errors?
    end

    def errors
      # Run validations
      has_errors?
      has_entry_errors?

      # Build errors
      all_errors = {}
      all_errors[:aba] = self.error_collection unless self.error_collection.empty?
      entry_error_collection = entries.each_with_index
        .map { |t, i| [i, t.error_collection] }
        .reject { |e| e[1].nil? || e[1].empty? }
        .to_h
      all_errors[:entries] = entry_error_collection unless entry_error_collection.empty?

      all_errors unless all_errors.empty?
    end

    private

    def add_entry(type, attrs)
      (attrs.instance_of?(type) ? attrs : type.new(attrs)).tap do |entry|
        entries << entry
      end
    end

    def has_entry_errors?
      entries.map { |t| t.valid? }.include?(false)
    end

    def descriptive_record
      # Record type
      # Max: 1
      # Char position: 1
      output = "0"

      # Optional branch number of the funds account with a hyphen in the 4th character position
      # Char position: 2-18
      # Max: 17
      # Blank filled
      output += self.bsb.nil? ? " " * 17 : self.bsb.to_s.ljust(17)

      # Sequence number
      # Char position: 19-20
      # Max: 2
      # Zero padded
      output += "01"

      # Name of user financial instituion
      # Max: 3
      # Char position: 21-23
      output += self.financial_institution.to_s

      # Reserved
      # Max: 7
      # Char position: 24-30
      output += " " * 7

      # Name of User supplying File
      # Char position: 31-56
      # Max: 26
      # Full BECS character set valid
      # Blank filled
      output += self.user_name.to_s.ljust(26)

      # Direct Entry User ID
      # Char position: 57-62
      # Max: 6
      # Zero padded
      output += self.user_id.to_s.rjust(6, "0")

      # Description of payments in the file (e.g. Payroll, Creditors etc.)
      # Char position: 63-74
      # Max: 12
      # Full BECS character set valid
      # Blank filled
      output += self.description.to_s.ljust(12)

      # Date on which the payment is to be processed
      # Char position: 75-80
      # Max: 6
      output += self.process_at.to_s.rjust(6, "0")

      # Reserved
      # Max: 40
      # Char position: 81-120
      output += " " * 40
    end

    def batch_control_record
      cached_credit_total_amount = credit_total_amount
      cached_debit_total_amount = debit_total_amount

      # Record type
      # Max: 1
      # Char position: 1
      output = "7"

      # BSB Format Filler
      # Max: 7
      # Char position: 2-8
      output += "999-999"

      # Reserved
      # Max: 12
      # Char position: 9-20
      output += " " * 12

      # Net total
      # Max: 10
      # Char position: 21-30
      output += (cached_credit_total_amount - cached_debit_total_amount).abs.to_s.rjust(10, "0")

      # Credit Total Amount
      # Max: 10
      # Char position: 31-40
      output += cached_credit_total_amount.to_s.rjust(10, "0")

      # Debit Total Amount
      # Max: 10
      # Char position: 41-50
      output += cached_debit_total_amount.to_s.rjust(10, "0")

      # Reserved
      # Max: 24
      # Char position: 51-74
      output += " " * 24

      # Total Item Count
      # Max: 6
      # Char position: 75-80
      output += entries.size.to_s.rjust(6, "0")

      # Reserved
      # Max: 40
      # Char position: 81-120
      output += " " * 40
    end

    def credit_total_amount
      credit_total_amount ||= entries.reject(&:debit?).sum(&method(:entry_amount))
    end

    def debit_total_amount
      entries.select(&:debit?).sum(&method(:entry_amount))
    end

    def entry_amount(entry)
      entry.amount.to_s[/\A\-?\d+\z/] ? Integer(entry.amount).abs : 0
    end
  end
end


lib/aba/return.rb
# frozen_string_literal: true

class Aba
  class Return < Entry
    include Aba::Validations

    attr_accessor :account_number, :transaction_code, :amount, :account_name,
                  :bsb, :trace_bsb, :trace_account_number, :name_of_remitter,
                  :return_code, :lodgement_reference,
                  :original_processing_day, :original_user_id

    # BSB
    validates_bsb               :bsb

    # Account Number
    validates_account_number    :account_number

    # Indicator
    validates_return_code       :return_code

    # Transaction Code
    validates_transaction_code  :transaction_code

    # Amount
    validates_integer           :amount

    # Original Day of Processing
    validates_integer           :original_processing_day, :unsigned

    # Original User Id
    validates_max_length        :original_user_id, 6
    validates_integer           :original_user_id, :unsigned

    # Account Name
    validates_max_length        :account_name, 32
    validates_becs              :account_name

    # Lodgement Reference
    validates_max_length        :lodgement_reference, 18
    validates_becs              :lodgement_reference

    # Trace Record
    validates_bsb               :trace_bsb
    validates_account_number    :trace_account_number

    # Name of Remitter
    validates_max_length        :name_of_remitter, 16
    validates_becs              :name_of_remitter

    # Allow dashes to be input, but remove them from output
    def account_number
      @account_number ? @account_number.to_s.delete('-') : nil
    end

    def to_s
      raise 'Transaction data is invalid - check the contents of `errors`' unless valid?

      format('2%-7s%9s%1d%2d%010d%-32s%-18s%-7s%9s%-16s%02d%6s',
             bsb,
             account_number,
             return_code,
             transaction_code,
             amount.to_i.abs,
             account_name,
             lodgement_reference,
             trace_bsb,
             trace_account_number,
             name_of_remitter,
             original_processing_day.to_i,
             original_user_id)
    end
  end
end


lib/aba/transaction.rb
# frozen_string_literal: true

class Aba
  class Transaction < Entry
    include Aba::Validations

    attr_accessor :account_number, :transaction_code, :amount, :account_name,
                  :bsb, :trace_bsb, :trace_account_number, :name_of_remitter,
                  :witholding_amount, :indicator, :lodgement_reference

    # BSB
    validates_bsb               :bsb

    # Account Number
    validates_account_number    :account_number

    # Indicator
    validates_indicator         :indicator

    # Transaction Code
    validates_transaction_code  :transaction_code

    # Amount
    validates_integer           :amount
    validates_max_length        :amount, 10

    # Account Name
    validates_max_length        :account_name, 32
    validates_becs              :account_name

    # Lodgement Reference
    validates_max_length        :lodgement_reference, 18
    validates_becs              :lodgement_reference

    # Trace Record
    validates_bsb               :trace_bsb
    validates_account_number    :trace_account_number

    # Name of Remitter
    validates_max_length        :name_of_remitter, 16
    validates_becs              :name_of_remitter

    # Allow dashes to be input, but remove them from output
    def account_number
      @account_number ? @account_number.to_s.gsub('-', '') : nil
    end

    # Sane default for majority of use cases as per BECS
    def indicator
      @indicator || ' '
    end

    # Optional as per BECS
    def lodgement_reference
      @lodgement_reference || ''
    end

    def to_s
      raise RuntimeError, 'Transaction data is invalid - check the contents of `errors`' unless valid?

      # Record type
      output = "1"

      # BSB of account
      output += bsb

      # Account number
      #raise RuntimeError, 'Transaction is missing account_number' unless account_number
      output += account_number.to_s.rjust(9, " ")

      # Withholding Tax Indicator
      # "N" – for new or varied Bank/State/Branch number or name details, otherwise blank filled.
      # "T" - for a drawing under a Transaction Negotiation Authority.
      # "W" – dividend paid to a resident of a country where a double tax agreement is in force.
      # "X" – dividend paid to a resident of any other country.
      # "Y" – interest paid to all non-residents.
      output += indicator.to_s.ljust(1, " ")

      # Transaction Code
      # 50 General Credit.
      # 53 Payroll.
      # 54 Pension.
      # 56 Dividend.
      # 57 Debenture Interest.
      # 13 General Debit.
      output += transaction_code.to_s

      # Amount to be credited or debited
      output += amount.to_i.abs.to_s.rjust(10, "0")

      # Title of Account
      # Full BECS character set valid
      output += account_name.to_s.ljust(32, " ")

      # Lodgement Reference Produced on the recipient’s Account Statement.
      # Full BECS character set valid
      output += lodgement_reference.to_s.ljust(18, " ")

      # Trace BSB Number
      output += trace_bsb

      # Trace Account Number
      output += trace_account_number.to_s.rjust(9, " ")

      # Name of Remitter Produced on the recipient’s Account Statement
      # Full BECS character set valid
      output += name_of_remitter.to_s.ljust(16, " ")

      # Withholding amount in cents
      output += (witholding_amount || 0).abs.to_s.rjust(8, "0")
    end
  end
end


lib/aba/validations.rb
# frozen_string_literal: true

class Aba
  module Validations
    attr_accessor :error_collection

    BECS_PATTERN = /\A[\w\+\-\@\ \$\!\%\&\(\)\*\.\/\#\=\:\;\?\,\'\[\]\_\^]*\Z/
    INDICATORS = [' ', 'N', 'T', 'W', 'X', 'Y']
    DEBIT_TRANSACTION_CODES = [13]
    CREDIT_TRANSACTION_CODES = [50, 51, 52, 53, 54, 55, 56, 57]

    def self.included(base)
      base.instance_eval do
        @_validations = {}
      end

      base.send :extend, ClassMethods
    end

    def self.transaction_codes
      DEBIT_TRANSACTION_CODES + CREDIT_TRANSACTION_CODES
    end

    def valid?
      !has_errors?
    end

    alias_method 'errors', 'error_collection'

    private

    # Run all validations
    def has_errors?
      self.error_collection = []

      self.class.instance_variable_get(:@_validations).each do |attribute, validations|
        value = send(attribute)

        validations.each do |type, param|
          case type
          when :presence
            self.error_collection << "#{attribute} is empty" if value.nil? || value.to_s.empty?
          when :bsb
            unless((param && value.nil?) || value =~ /^\d{3}-\d{3}$/)
              self.error_collection << "#{attribute} format is incorrect"
            end
          when :max_length
            self.error_collection << "#{attribute} length must not exceed #{param} characters" if value.to_s.length > param
          when :length
            self.error_collection << "#{attribute} length must be exactly #{param} characters" if value.to_s.length != param
          when :integer
            if param
              self.error_collection << "#{attribute} must be a number" unless value.to_s =~ /\A[+-]?\d+\Z/
            else
              self.error_collection << "#{attribute} must be an unsigned number" unless value.to_s =~ /\A\d+\Z/
            end
          when :account_number
            if value.to_s =~ /\A[0\ ]+\Z/ || value.to_s !~ /\A[a-z\d\ ]{1,9}\Z/i
              self.error_collection << "#{attribute} must be a valid account number"
            end
          when :becs
            self.error_collection << "#{attribute} must not contain invalid characters" unless value.to_s =~ BECS_PATTERN
          when :indicator
            list = INDICATORS.join('\', \'')
            self.error_collection << "#{attribute} must be a one of '#{list}'" unless INDICATORS.include?(value.to_s)
          when :transaction_code
            self.error_collection << "#{attribute} must be one of #{Validations.transaction_codes.join(', ')}" unless Validations.transaction_codes.include?(value.to_i)
          end
        end
      end

      !self.error_collection.empty?
    end

    module ClassMethods
      def validates_presence_of(*attributes)
        attributes.each do |a|
          add_validation_attribute(a, :presence)
        end
      end

      def validates_bsb(attribute, options = {})
        options[:allow_blank] ||= false
        add_validation_attribute(attribute, :bsb, options[:allow_blank])
      end

      def validates_max_length(attribute, length)
        add_validation_attribute(attribute, :max_length, length)
      end

      def validates_length(attribute, length)
        add_validation_attribute(attribute, :length, length)
      end

      def validates_integer(attribute, signed = true)
        add_validation_attribute(attribute, :integer, signed)
      end

      def validates_account_number(attribute)
        add_validation_attribute(attribute, :account_number)
      end

      def validates_becs(attribute)
        add_validation_attribute(attribute, :becs)
      end

      def validates_indicator(attribute)
        add_validation_attribute(attribute, :indicator)
      end

      def validates_transaction_code(attribute)
        add_validation_attribute(attribute, :transaction_code)
      end

      def validates_return_code(attribute)
        add_validation_attribute(attribute, :integer, :return_code)
      end

      private

      def add_validation_attribute(attribute, type, param = true)
        @_validations[attribute] = {} unless @_validations[attribute]
        @_validations[attribute][type] = param
      end
    end
  end
end
